core/unit_test/TestMathematicalFunctions.hpp

```diff
@@ -1837,6 +1841,127 @@ TEST(TEST_CATEGORY, mathematical_functions_impl_half_fallback) {
+    }
+
+    // Equality (from==toward) Handling
+    if (nextafter(pos_one, pos_one) != pos_one ||
+        nextafter(pos_zero, pos_zero) != pos_zero ||
+        nextafter(neg_zero, neg_zero) != neg_zero ||
+        nextafter(pos_inf, pos_inf) != pos_inf ||
+        nextafter(neg_inf, neg_inf) != neg_inf) {
+      ++e;
+      Kokkos::printf("failed half precision nextafter(equality)\n");
+    }
+
+    // Zero Handling
+    if (nextafter(pos_zero, pos_one) != pos_smallest ||
+        nextafter(pos_zero, neg_one) != neg_smallest ||
+        nextafter(pos_zero, neg_zero) != neg_zero ||
+        nextafter(neg_zero, pos_one) != pos_smallest ||
+        nextafter(neg_zero, neg_one) != neg_smallest ||
+        nextafter(neg_zero, pos_zero) != pos_zero) {
+      ++e;
+      Kokkos::printf("failed half precision nextafter(zero)\n");
+    }
+
+    // From Negative Non Zero Handling
+    const FP16Type after_neg_one = Kokkos::bit_cast<FP16Type>(
+        std::uint16_t(Kokkos::bit_cast<std::uint16_t>(neg_one) - 1));
+    const FP16Type before_neg_one = Kokkos::bit_cast<FP16Type>(
+        std::uint16_t(Kokkos::bit_cast<std::uint16_t>(neg_one) + 1));
+    if (nextafter(neg_smallest, pos_zero) != neg_zero ||
+        nextafter(neg_one, pos_one) != after_neg_one ||
+        nextafter(neg_one, neg_two) != before_neg_one ||
+        nextafter(neg_max, neg_inf) != neg_inf) {
+      ++e;
+      Kokkos::printf("failed half precision nextafter(negative)\n");
+    }
+
+    // From Positive Non Zero Handling
+    const FP16Type after_pos_one = Kokkos::bit_cast<FP16Type>(
+        std::uint16_t(Kokkos::bit_cast<std::uint16_t>(pos_one) + 1));
+    const FP16Type before_pos_one = Kokkos::bit_cast<FP16Type>(
+        std::uint16_t(Kokkos::bit_cast<std::uint16_t>(pos_one) - 1));
+    if (nextafter(pos_smallest, neg_zero) != pos_zero ||
+        nextafter(pos_one, neg_one) != before_pos_one ||
+        nextafter(pos_one, pos_two) != after_pos_one ||
+        nextafter(pos_max, pos_inf) != pos_inf) {
+      ++e;
+      Kokkos::printf("failed half precision nextafter(positive)\n");
+    }
+
+    // From Inf Handling
+    if (nextafter(pos_inf, pos_one) != pos_max ||
+        nextafter(neg_inf, neg_one) != neg_max ||
+        nextafter(pos_inf, pos_inf) != pos_inf ||
+        nextafter(neg_inf, neg_inf) != neg_inf) {
+      ++e;
+      Kokkos::printf("failed half precision nextafter(inf)\n");
+    }
+  }
+};
+
+TEST(TEST_CATEGORY, mathematical_functions_nextafter_fp16) {
+#if !(defined(KOKKOS_ENABLE_CUDA) && defined(KOKKOS_COMPILER_MSVC))
+#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
+  TestNextAfterHalf<TEST_EXECSPACE, Kokkos::Experimental::half_t>();
+#endif
+#if defined(KOKKOS_BHALF_T_IS_FLOAT) && !KOKKOS_BHALF_T_IS_FLOAT
+  TestNextAfterHalf<TEST_EXECSPACE, Kokkos::Experimental::bhalf_t>();
+#endif
+#else
+  GTEST_SKIP() << "FIXME MSVC nextafter for half precision "
+                  "not implemented yet";
+#endif
```

Comment:
Readability
```suggestion
#if defined(KOKKOS_ENABLE_CUDA) && defined(KOKKOS_COMPILER_MSVC)
  GTEST_SKIP() << "FIXME MSVC nextafter for half precision "
                  "not implemented yet";
#endif

bool skipped = true;
#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
  skipped = false;
  TestNextAfterHalf<TEST_EXECSPACE, Kokkos::Experimental::half_t>();
#endif
#if defined(KOKKOS_BHALF_T_IS_FLOAT) && !KOKKOS_BHALF_T_IS_FLOAT
  skipped = false;
  TestNextAfterHalf<TEST_EXECSPACE, Kokkos::Experimental::bhalf_t>();
#endif

if (skipped)
  GTEST_SKIP() << "no 16-bit floating-point precision support";
```

I think we should skip if neither

Reply:
Sure, that is better

Reply:
Actually, this does not work. 
In MSVC, half is there but nextafter for that is not implemented.

I think we need something like

```C++
#if defined(KOKKOS_ENABLE_CUDA) && defined(KOKKOS_COMPILER_MSVC)
  GTEST_SKIP() << "FIXME MSVC nextafter for half precision "
                  "not implemented yet";
#else
bool skipped = true;
#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
  skipped = false;
  TestNextAfterHalf<TEST_EXECSPACE, Kokkos::Experimental::half_t>();
#endif
#if defined(KOKKOS_BHALF_T_IS_FLOAT) && !KOKKOS_BHALF_T_IS_FLOAT
  skipped = false;
  TestNextAfterHalf<TEST_EXECSPACE, Kokkos::Experimental::bhalf_t>();
#endif
if (skipped)
  GTEST_SKIP() << "no 16-bit floating-point precision support";
#endif
```